client/client.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h> 
+#include <error.h>
+#include <argp.h>
+
+static const uint16_t MSG_MAX_LEN = (uint16_t)~0U;
+
+/* Argp's global variables. */
+const char *argp_program_version =
+	"Gatekeeper dynamic configuration client 1.0";
+
+/* Arguments. */
+static char adoc[] = "<PATH>";
+
+static char doc[] = "Gatekeeper Client -- configure the Gatekeeper via "
+	"dynamic configuration";
+
+static struct argp_option options[] = {
+	{"server-path",		's',	"FILE",		0,
+		"Path to create the UNIX socket",	1},
+	{ 0 }
+};
+
+struct args {
+	const char *server_path;
+	const char *lua_script_path;
+};
+
+static error_t
+parse_opt(int key, char *arg, struct argp_state *state)
+{
+	struct args *args = state->input;
+
+	switch (key) {
+	case 's':
+		args->server_path = arg;
+		break;
+
+	case ARGP_KEY_INIT:
+		args->lua_script_path = NULL;
+		break;
+
+	case ARGP_KEY_ARG:
+		if (args->lua_script_path) {
+			argp_error(state,
+				"Wrong number of arguments; only one is allowed");
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	FILE *fp;
+	char *line = NULL;
+	char file_str[MSG_MAX_LEN];
+	char send_buff[MSG_MAX_LEN];
+	char recv_buff[MSG_MAX_LEN];
```

Comment:
Add two bytes to `send_buff` and `recv_buff`, this way you can set/receive the size of the buffer at the beginning. I think this will simplify the code a bit because you can send/receive everything at once instead of making at least two calls.

Reply:
According to our protocol in dynamic configuration, we will call the receive/send function twice to get the #bytes and the real message respectively (https://github.com/AltraMayor/gatekeeper/blob/master/config/dynamic.c#L233). If we combine the two messages into one, then the dynamic configuration will get stuck.

Reply:
It'll not get stuck because the first `read_nbytes()` only reads two bytes.

Reply:
Thanks! You are right.